x-pack/plugins/security_solution/public/common/containers/use_search_strategy/index.tsx

```diff
@@ -0,0 +1,175 @@
+    timerange: transformTimerange,
+  } = getTransformChangesIfTheyExist({
+    factoryQueryType,
+    indices: defaultIndex,
+    filterQuery,
+    timerange,
+  });
+
+  return data.search.search<RequestBasicOptions, ResponseType>(
+    {
+      ...requestProps,
+      factoryQueryType: transformFactoryQueryType,
+      defaultIndex: transformIndices,
+      timerange: transformTimerange,
+      filterQuery,
+    },
+    {
+      strategy: 'securitySolutionSearchStrategy',
+      abortSignal: signal,
+    }
+  );
+};
+
+const searchComplete = <ResponseType extends IKibanaSearchResponse>(
+  props: UseSearchStrategyRequestArgs
+): Observable<ResponseType> => {
+  return search<ResponseType>(props).pipe(
+    filter((response) => {
+      return isErrorResponse(response) || isCompleteResponse(response);
+    })
+  );
+};
+
+export const useSearchStrategy = <QueryType extends FactoryQueryTypes>({
+  factoryQueryType,
+  initialResult,
+  errorMessage,
+}: {
+  factoryQueryType: QueryType;
+  /**
+   * `result` initial value. It is used until the search strategy returns some data.
+   */
+  initialResult: Omit<StrategyResponseType<QueryType>, 'rawResponse'>;
+  /**
+   * Message displayed to the user on a Toast when an erro happens.
+   */
+  errorMessage?: string;
+}) => {
+  const abortCtrl = useRef(new AbortController());
+  const { getTransformChangesIfTheyExist } = useTransforms();
+
+  const refetch = useRef<inputsModel.Refetch>(noop);
+  const { data } = useKibana().services;
+  const { addError } = useAppToasts();
+
+  const { start, error, result, loading } = useObservable<
+    [UseSearchStrategyRequestArgs],
+    StrategyResponseType<QueryType>
+  >(searchComplete);
+
+  useEffect(() => {
+    if (error != null) {
+      addError(error, {
+        title: errorMessage ?? i18n.DEFAULT_ERROR_SEARCH_STRATEGY(factoryQueryType),
+      });
+    }
+  }, [addError, error, errorMessage, factoryQueryType]);
+
+  const searchCb = useCallback(
+    (props: OptionalSignalArgs<StrategyRequestType<QueryType>>) => {
+      const asyncSearch = () => {
+        abortCtrl.current = new AbortController();
+        start({
+          ...props,
+          data,
+          factoryQueryType,
+          getTransformChangesIfTheyExist,
+          signal: abortCtrl.current.signal,
+        } as never); // This typescast is required because every StrategyRequestType instance has different fields.
+      };
+
+      abortCtrl.current.abort();
+      asyncSearch();
+
+      refetch.current = asyncSearch;
+    },
+    [data, start, factoryQueryType, getTransformChangesIfTheyExist]
+  );
+
+  useEffect(() => {
+    return () => {
+      abortCtrl.current.abort();
+    };
+  }, []);
+
+  const [formatedResult, inspect] = useMemo(
+    () => [
+      omit<StrategyResponseType<QueryType>, 'rawResponse'>('rawResponse', result),
+      getInspectResponse(result, {
+        dsl: [],
+        response: [],
+      }),
+    ],
+    [result]
+  );
+
+  return {
```

Comment:
Good question! Let me go over each property.

* `id` is just a constant defined inside the container. We can move its declaration to the component calling `useSearchStrategy`

* `isInspected` is a mystery to me. On every container that I checked, it is hardcoded as false. 
  
  ![Mar-16-2022 10-38-19](https://user-images.githubusercontent.com/1490444/158561040-2aceed87-4cb6-444d-9a88-c9ebad729cfe.gif)
  
  Do we need it? `¯\_(ツ)_/¯`
  

* `pageInfo` and `totalCount` are properties returned by the API and it is included in the response.
  ```js
  setNetworkTopCountriesResponse((prevResponse) => ({
     …
     pageInfo: response.pageInfo,                 
     totalCount: response.totalCount,
     …
  }));
  ```

* `loadPage` is a call-back function that updates the request pagination property. It is called by PaginatedTable at the same time it calls `updateActivePage`. I think that we don't need both of them. We could set `loadPage={noop}` like I did on `UserRiskScoreQueryTabBody`. And only define `updateActivePage`
  ```js
   const updateActivePage = useCallback(
     (newPage) =>
       dispatch(
         reduxActions.updateTableActivePage({
           activePage: newPage,
           …
         })
       ),
     [type, dispatch]
   );
  ```
  That would update the store, and then we can get the pagination data from the store and set it as a parameter to the hook, like on here:
  ```js
   const [loading, { data, totalCount, inspect, isInspected, refetch }] = useHostRiskScore({
      …
     pagination,
   });
  ```
  
If we think it would be too much work to change how every pagination works, we could create a custom hook named `usePaginatedSearchStrategy` that mimics the current container API. 

The safer strategy might be to focus on the most straightforward API calls first, like fetch host and network details and some visualizations. If `useSearchStrategy`works well for these scenarios, we can migrate the tables.

Reply:
ok that sounds good to me

Reply:
It would be nice to open a follow up issue for the migrating existing code to use this hook.
Thank you for the initiative to simplify the code base!